### PR TITLE
Java module

### PR DESCRIPTION
**Pull request by tristan957**
Java module will support Java development within Meson just like the Python module does for Python. First order of business is generating native header files for use in JNI development.

Not really sure what else will go in this module, but I think moving the `jar` method under here could be preferable. Other JVM languages can create JAR files, but Meson's jar() assumes Java. Could be good to scope the naked jar() to be Java-specific.

### CODE REVIEW

File: mesonbuild/modules/java.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+class JavaModule(ExtensionModule):
+    @FeatureNew('Java Module', '0.59.0')
+    def __init__(self, interpreter: Interpreter):
+        super().__init__(interpreter)
+        self.methods.update({
+            'generate_native_header': self.generate_native_header,
+        })
+
+    @typed_pos_args('generate_native_header', (str, FileHolder), min_varargs=1, max_varargs=1)
+    # @permitted_kwargs({'package'})
```
Comment (tristan957) [inline mesonbuild/modules/java.py:35]: I do not understand why this doesn't work. Help is appreciated. When uncommented...

```
Traceback (most recent call last):
  File "/home/tristan957/Projects/meson/mesonbuild/mesonmain.py", line 223, in run
    return options.run_func(options)
  File "/home/tristan957/Projects/meson/mesonbuild/msetup.py", line 281, in run
    app.generate()
  File "/home/tristan957/Projects/meson/mesonbuild/msetup.py", line 184, in generate
    self._generate(env)
  File "/home/tristan957/Projects/meson/mesonbuild/msetup.py", line 223, in _generate
    intr.run()
  File "/home/tristan957/Projects/meson/mesonbuild/interpreter/interpreter.py", line 2424, in run
    super().run()
  File "/home/tristan957/Projects/meson/mesonbuild/interpreterbase/interpreterbase.py", line 132, in run
    self.evaluate_codeblock(self.ast, start=1)
  File "/home/tristan957/Projects/meson/mesonbuild/interpreterbase/interpreterbase.py", line 157, in evaluate_codeblock
    raise e
  File "/home/tristan957/Projects/meson/mesonbuild/interpreterbase/interpreterbase.py", line 150, in evaluate_codeblock
    self.evaluate_statement(cur)
  File "/home/tristan957/Projects/meson/mesonbuild/interpreterbase/interpreterbase.py", line 163, in evaluate_statement
    return self.function_call(cur)
  File "/home/tristan957/Projects/meson/mesonbuild/interpreterbase/interpreterbase.py", line 549, in function_call
    res = func(node, func_args, kwargs)
  File "/home/tristan957/Projects/meson/mesonbuild/interpreterbase/decorators.py", line 598, in wrapped
    return f(*wrapped_args, **wrapped_kwargs)
  File "/home/tristan957/Projects/meson/mesonbuild/interpreterbase/decorators.py", line 102, in wrapped
    return f(*wrapped_args, **wrapped_kwargs)
  File "/home/tristan957/Projects/meson/mesonbuild/interpreter/interpreter.py", line 1909, in func_subdir
    self.evaluate_codeblock(codeblock)
  File "/home/tristan957/Projects/meson/mesonbuild/interpreterbase/interpreterbase.py", line 157, in evaluate_codeblock
    raise e
  File "/home/tristan957/Projects/meson/mesonbuild/interpreterbase/interpreterbase.py", line 150, in evaluate_codeblock
    self.evaluate_statement(cur)
  File "/home/tristan957/Projects/meson/mesonbuild/interpreterbase/interpreterbase.py", line 165, in evaluate_statement
    self.assignment(cur)
  File "/home/tristan957/Projects/meson/mesonbuild/interpreterbase/interpreterbase.py", line 910, in assignment
    value = self.evaluate_statement(node.value)
  File "/home/tristan957/Projects/meson/mesonbuild/interpreterbase/interpreterbase.py", line 167, in evaluate_statement
    return self.method_call(cur)
  File "/home/tristan957/Projects/meson/mesonbuild/interpreterbase/interpreterbase.py", line 593, in method_call
    return self._holderify(obj.method_call(method_name, args, kwargs))
  File "/home/tristan957/Projects/meson/mesonbuild/interpreter/interpreterobjects.py", line 769, in method_call
    ret = method(state, args, kwargs)
TypeError: _inner() takes 3 positional arguments but 4 were given
```

File: mesonbuild/modules/java.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+class JavaModule(ExtensionModule):
+    @FeatureNew('Java Module', '0.59.0')
+    def __init__(self, interpreter: Interpreter):
+        super().__init__(interpreter)
+        self.methods.update({
+            'generate_native_header': self.generate_native_header,
+        })
+
+    @typed_pos_args('generate_native_header', (str, FileHolder), min_varargs=1, max_varargs=1)
```
Comment (tristan957) [inline mesonbuild/modules/java.py:34]: @dcbaker would appreciate reviewing this specifically.

File: mesonbuild/modules/java.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        else:
+            header = f'{pathlib.Path(file.fname).stem}.h'
+
+        ct_kwargs = {
+            'input': file,
+            'output': header,
+            'command': [
+                *javac.get_command(),
+                '-d',
+                os.path.join(state.backend.build_dir, state.subdir),
```
Comment (tristan957) [inline mesonbuild/modules/java.py:55]: Do we have some sort of trash directory or private dir that I could throw the .class file in?

**Comment by tristan957** [comment]
I'm also not sure what custom target kwargs I should support of any. Headers that you generate with this aren't supposed to be installed best I can tell.

File: mesonbuild/modules/java.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
+# Copyright 2021 The Meson development team
```
Comment (tristan957) [inline mesonbuild/modules/java.py:1]: I am also not sure how copyright works here. Do I put my name or leave as is?

File: mesonbuild/modules/java.py (hunk 1 above)
Comment (tristan957) [reply mesonbuild/modules/java.py:35]: I don't understand how 4 arguments were passed though.

File: mesonbuild/modules/java.py (hunk 2 above)
Comment (dcbaker) [reply mesonbuild/modules/java.py:34]: You don't need the min_ and max_ varargs here, your function only uses fixed positional arguments, varargs (variadic arguments) are for functions like `file()` which takes 0 or more strings and returns the an array with the same number of File elements. I have no idea what happens if you set those flags and don't set `vaargs=...`, you could be running into an odd corner of that.

File: mesonbuild/modules/java.py (hunk 3 above)
Comment (dcbaker) [reply mesonbuild/modules/java.py:55]: @PRIVATE_DIR@

File: mesonbuild/modules/java.py (hunk 4 above)
Comment (dcbaker) [reply mesonbuild/modules/java.py:1]: We don't have a hard and fast rule. When I'm developing on my own time I use `The Meson development team`, when I'm on work time my employer wants me to put `Copyright © 2021 Intel Corporation`, so I do that, and no one has complained

File: mesonbuild/modules/java.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+import pathlib
+from mesonbuild.mesonlib.universal import File
+from mesonbuild.build import CustomTarget
+from mesonbuild.optinterpreter import permitted_kwargs
+import os
+from mesonbuild.interpreterbase.decorators import FeatureNew, typed_pos_args
+import typing as T
+from . import ExtensionModule, ModuleReturnValue, ModuleState
+from ..interpreter import Interpreter
+from mesonbuild.interpreter.interpreterobjects import FileHolder
```
Comment (dcbaker) [inline mesonbuild/modules/java.py:24]: The usual way that python imports are sorted are:
```
<builtin package 1>
<builtin package 2>
...

<third party packge 1>
...

<local package/module 1>
<local package/module 2>
...
```
and alphabeitcally within thant

File: mesonbuild/modules/java.py [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+    @typed_pos_args('generate_native_header', (str, FileHolder), min_varargs=1, max_varargs=1)
+    # @permitted_kwargs({'package'})
+    def generate_native_header(self, state: ModuleState, args: T.Tuple[T.Union[str, FileHolder]], kwargs: T.Dict[str, str]) -> ModuleReturnValue:
+        assert state.backend
+
+        javac = state.find_program("javac")
+        package = kwargs.get('package', None)
+
+        file = args[0].held_object if isinstance(args[0], FileHolder) else File(False, state.subdir, args[0])
```
Comment (dcbaker) [inline mesonbuild/modules/java.py:42]: There's a method in the interpreter for this

File: mesonbuild/modules/java.py (hunk 4 above)
Comment (tristan957) [reply mesonbuild/modules/java.py:1]: I will just leave it as is then! Just did it on my own time.

File: mesonbuild/modules/java.py (hunk 6 above)
Comment (tristan957) [reply mesonbuild/modules/java.py:42]: If you remember what it is let me know!

Comment (dcbaker) [reply mesonbuild/modules/java.py:42]: source_strings_to_files.

You may need to add:
```python
T.overload
def source_strings_to_files(self, sources: T.Iterable['mesonlib.FileOrString']) -> T.List[File]: ...
```

to the interpreter to get the typing right. I have that patch in one of my branches somewhere...

File: mesonbuild/modules/java.py (hunk 1 above)
Comment (dcbaker) [reply mesonbuild/modules/java.py:35]: You don't want `permitted_kwargs` anway, that's old.
```python
@typed_kwargs(
  'java.generate_native_header',
  KwargInfo('package', (str, None)),
)
def ...:
   package = kwargs['package']
```

Comment (dcbaker) [reply mesonbuild/modules/java.py:35]: Sorry, I just noticed that you were using permitted_kwargs...

File: mesonbuild/modules/java.py (hunk 6 above)
Comment (tristan957) [reply mesonbuild/modules/java.py:42]: I'll leave the typing alone unless you want me to add it here if you can't remember where the patch is.

**Comment by tristan957** [comment]
So I am using @PRIVATE_DIR@, and tests pass for everything but Arch. Am I doing something wrong?

```
[1/6] /usr/sbin/javac -d '/__w/meson/meson/b 458a3d4a57/com_mesonbuild_JdkTest.h.p' -h '/__w/meson/meson/b 458a3d4a57/src/com/mesonbuild' '../test cases/java/9 jdk/src/com/mesonbuild/JdkTest.java'
FAILED: /__w/meson/meson/b 458a3d4a57/com_mesonbuild_JdkTest.h 
/usr/sbin/javac -d '/__w/meson/meson/b 458a3d4a57/com_mesonbuild_JdkTest.h.p' -h '/__w/meson/meson/b 458a3d4a57/src/com/mesonbuild' '../test cases/java/9 jdk/src/com/mesonbuild/JdkTest.java'
javac: directory not found: /__w/meson/meson/b 458a3d4a57/com_mesonbuild_JdkTest.h.p
```

**Comment by dcbaker** [comment]
it's odd that path is absolute, is that really correct?

**Comment by tristan957** [comment]
The other oddity here is that the private dir gets created at the root of the build tree. No idea why that is happening.

File: mesonbuild/modules/java.py [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+                *javac.get_command(),
+                '-d',
+                '@PRIVATE_DIR@',
+                '-h',
+                os.path.join(state.backend.build_dir, state.subdir),
+                '@INPUT@',
+            ]
+        }
+
+        target = CustomTarget(os.path.basename(header), state.backend.build_dir, state.subproject, backend=state.backend, kwargs=ct_kwargs)
```
Comment (jpakkane) [inline mesonbuild/modules/java.py:62]: The second argument should probably be `state.subdir` rather than `state.backend_build_dir`.

**Comment by tristan957** [comment]
```
 [1/6] /usr/sbin/javac -d src/com/mesonbuild/com_mesonbuild_JdkTest.h.p -h '/__w/meson/meson/b 458a3d4a57/src/com/mesonbuild' '../test cases/java/9 jdk/src/com/mesonbuild/JdkTest.java'
FAILED: src/com/mesonbuild/com_mesonbuild_JdkTest.h 
/usr/sbin/javac -d src/com/mesonbuild/com_mesonbuild_JdkTest.h.p -h '/__w/meson/meson/b 458a3d4a57/src/com/mesonbuild' '../test cases/java/9 jdk/src/com/mesonbuild/JdkTest.java'
javac: directory not found: src/com/mesonbuild/com_mesonbuild_JdkTest.h.p
```
Same issue, but the private dir is now definitely in the right place, which is good.

**Comment by tristan957** [comment]
Just pushed what I believe to be the final version (for now).

File: docs/markdown/snippets/java_module.md [hunk 8]
```diff
@@ -0,0 +1,9 @@
+## Java Module
+
+The Java module has been added to Meson. The Java module allows users to
+generate native header files without needing to use a `custom_target()`.
+
+```meson
+jmod = import('java')
+native_header = jmod.generate_native_header('File.java', package: 'com.mesonbuild')
```
Comment (jpakkane) [inline docs/markdown/snippets/java_module.md:8]: This should also show how to pass `native_header` to a Java target.

Comment (tristan957) [reply docs/markdown/snippets/java_module.md:9]: Please let me know if the update is enough. I'm on my phone and away from keyboard atm.

**Comment by jpakkane** [comment]
```
ninja explain: deps for 'lib/libjdkjava.so.p/native.c.o' are missing
ninja explain: output src/com/mesonbuild/com_mesonbuild_JdkTest.h doesn't exist
ninja explain: lib/libjdkjava.so.p/native.c.o is dirty
ninja explain: deps for 'lib/libjdkjava.so.p/com_mesonbuild_JdkTest.c.o' are missing
ninja explain: lib/libjdkjava.so.p/com_mesonbuild_JdkTest.c.o is dirty
ninja explain: lib/libjdkjava.so is dirty
```

**Comment by jpakkane** [comment]
Releasing rc1 without this. If this gets fixed soon, we can consider adding it to rc2.

**Comment by tristan957** [comment]
I don't really understand the issue. If I run the test case on my machine it works great.

**Comment by tristan957** [comment]
Still at a loss as to why the Arch Linux check is failing. @eli-schwartz reports it works on his system although he has javac 16. I checked with Java 1.8.0, 11, and 16 on my Fedora 34 and it worked swimmingly. At this point, I am going to chalk it up to a CI issue because I am not sure what else I can do.

I rm -rfed the build dir everytime I switched java versions, so there was nothing stale on my system

**Comment by tristan957** [comment]
The same issue seems to be happening with `vs2019 vc2019x64ninja`. In what cases would the private dir not exist. It doesn't make any sense.

```
[1/6] "C:\hostedtoolcache\windows\Java_Adopt_jdk\8.0.292-10\x64\bin\javac.EXE" "-d" "src/com/mesonbuild/com_mesonbuild_JdkTest.h.p" "-h" "src/com/mesonbuild" "../test cases/java/9 jdk/src/com/mesonbuild/JdkTest.java"
FAILED: src/com/mesonbuild/com_mesonbuild_JdkTest.h 
"C:\hostedtoolcache\windows\Java_Adopt_jdk\8.0.292-10\x64\bin\javac.EXE" "-d" "src/com/mesonbuild/com_mesonbuild_JdkTest.h.p" "-h" "src/com/mesonbuild" "../test cases/java/9 jdk/src/com/mesonbuild/JdkTest.java"
javac: directory not found: src/com/mesonbuild/com_mesonbuild_JdkTest.h.p
Usage: javac <options> <source files>
use -help for a list of possible options
[2/6] "javac" "-g" "-d" "src\jdkjar.jar.p" "-s" "src\jdkjar.jar.p" "-sourcepath" "D:\a\1\s\test cases\java\9 jdk\src;D:\a\1\s\b 458a3d4a57\src;D:\a\1\s\b 458a3d4a57" "../test cases/java/9 jdk/src/com/mesonbuild/JdkTest.java"
ninja: build stopped: subcommand failed.
```

Am I missing something obvious?

**Comment by jpakkane** [comment]
> javac: directory not found: src/com/mesonbuild/com_mesonbuild_JdkTest.h.p

According to `javac` command line help, the output directory must already exist, it will not be created automatically and thus leads to failure.

**Comment by eli-schwartz** [comment]
I do not see that in the command line help.

Be that as it may, I've re-checked this test case after installing more Java versions.

```
$ sudo archlinux-java set java-16-openjdk && ./run_single_test.py "test cases/java/9 jdk"
[...]
java   : [unknown]  javac (unknown 16.0.2)
[...]
PASS: java: 9 jdk    (9 jdk)

$ sudo archlinux-java set java-8-openjdk && ./run_single_test.py "test cases/java/9 jdk"
[...]
java   : [unknown]  javac (unknown 1.8.0)
[...]
FAIL: java: 9 jdk    (9 jdk)
reason: Compiling source code failed.
The Meson build system
Version: 0.59.99
Source dir: /home/eschwartz/git/meson/test cases/java/9 jdk
Build dir: /tmp/b 8bfdc37f46
Build type: native build
Project name: jdkjava
Project version: undefined
C compiler for the host machine: ccache cc (gcc 11.1.0 "cc (GCC) 11.1.0")
C linker for the host machine: cc ld.bfd 2.36.1
Java compiler for the host machine: javac (unknown 1.8.0)
Host machine cpu family: x86_64
Host machine cpu: x86_64
Program java found: YES (/usr/bin/java)
Run-time dependency jdk found: YES 1.8.0
Program javac found: YES (/usr/bin/javac)
Build targets in project: 3

Found ninja-1.10.2 at /usr/bin/ninja
[1/6] Generating com_mesonbuild_JdkTest.h with a custom command
FAILED: src/com/mesonbuild/com_mesonbuild_JdkTest.h 
/usr/bin/javac -d src/com/mesonbuild/com_mesonbuild_JdkTest.h.p -h src/com/mesonbuild '../../home/eschwartz/git/meson/test cases/java/9 jdk/src/com/mesonbuild/JdkTest.java'
javac: directory not found: src/com/mesonbuild/com_mesonbuild_JdkTest.h.p
Usage: javac <options> <source files>
use -help for a list of possible options
[2/6] Compiling Java object '../../home/eschwartz/git/meson/test cases/java/9 jdk/src/com/mesonbuild/JdkTest.java'
ninja: build stopped: subcommand failed.
```

I can absolutely confirm that this is a ***behavior change*** in javac, at least on Arch Linux, and I don't know why on Fedora it would work differently.

If meson successfully configures with java 1.8 then I do believe it needs to create that directory.

**Comment by eli-schwartz** [comment]
Note that the PRIVATE_DIR appears to be created automatically for build.BuildTarget, but not for build.CustomTarget; the former is only used internally, I guess, while the latter is "the directory which custom commands are allowed to use internally, but who knows how". Apparently that means "you get to make sure the directory exists too"?

**Comment by dcbaker** [comment]
I noticed you didn't add the module to the `run_mypy` script, but it looks like it's annotated?

**Comment by tristan957** [comment]
Honestly I just didn't know how the type checking worked in this project. Will add it in!

**Comment by jpakkane** [comment]
Woop! All green, finally! \o/

LGTM, anyone else have comments?

**Comment by tristan957** [comment]
Sorry it took so long. Hopefully Java 8 is the only workaround we have to have. Maybe if someone requests less than 8 I can also come in and fix it.

**Review by dcbaker — APPROVED** [review]

File: docs/theme/extra/templates/navbar_links.html [hunk 9]
```diff
@@ -15,6 +15,7 @@
             ("Hotdoc-module.html","Hotdoc"), \
             ("i18n-module.html","i18n"), \
             ("Icestorm-module.html","Icestorm"), \
+            ("Java-module.html","Java") \
```
Comment (eli-schwartz) [inline docs/theme/extra/templates/navbar_links.html:18]: Oops, fatal missing comma

Comment (tristan957) [reply docs/theme/extra/templates/navbar_links.html:18]: Will submit a follow-up

Comment (tristan957) [reply docs/theme/extra/templates/navbar_links.html:18]: Oh you already did. Thanks.

**Pull request merged**